core/src/main/java/jenkins/slaves/JnlpSlaveAgentProtocol3.java

```diff
@@ -0,0 +1,112 @@
+package jenkins.slaves;
+
+import hudson.AbortException;
+import hudson.Extension;
+import hudson.model.Computer;
+import hudson.remoting.Channel;
+import hudson.remoting.ChannelBuilder;
+import hudson.slaves.SlaveComputer;
+import jenkins.AgentProtocol;
+import jenkins.model.Jenkins;
+import jenkins.security.ChannelConfigurator;
+import org.jenkinsci.remoting.engine.JnlpServer3Handshake;
+import org.jenkinsci.remoting.nio.NioChannelHub;
+
+import javax.inject.Inject;
+import java.io.IOException;
+import java.io.OutputStream;
+import java.io.PrintWriter;
+import java.net.Socket;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+/**
+ * Master-side implementation for JNLP3-connect protocol.
+ *
+ * <p>@see {@link org.jenkinsci.remoting.engine.JnlpProtocol3} for more details.
+ *
+ * @author Akshay Dayal
+ * @since 1.XXX
+ */
+@Extension
```

Comment:
Initial manual testing looks good - the JNLP node connects with JNLP3 protocol and I can build a project on the node.

I think more stress testing needs to be done before JNLP3 is made the default protocol. In this initial release JNLP3 should only be available if the user explicitly asks for it. There are 3 approaches:

Master-side control
    1. Making master support JNLP3 through a flag
    2. Making master support JNLP3 through UI option
Node-side control
    3. Making node try JNLP3 first using flag

Option 1 can be done in "easy" ways I suppose, this class can check for a boolean system property, if its not set fail and let JNLP2 take over.
Option 2 is a lot of work and I would not favor it.
Option 3 is easy, but requires some extra work for users starting up nodes with JNLP3.

I'm ok with either option 1 or 3, prefer 1 slightly since I like controlling the behavior from the master side, not the node.